### PR TITLE
state/remote/atlas: handle conflicts on equivalent states

### PR DESCRIPTION
**Pull request by phinze**
Atlas returns an HTTP 409 - Conflict if the pushed state reports the same
Serial number but the checksum of the raw content differs. This can
sometimes happen when Terraform changes state representation internally
between versions in a way that's semantically neutral but affects the JSON
output and therefore the checksum.

Here we detect and handle this situation by ticking the serial and retrying
iff for the previous state and the proposed state:

  * the serials match
  * the parsed states are Equal (semantically equivalent)

In other words, in this situation Terraform can override Atlas's detected
conflict by asserting that the state it is pushing is indeed correct.

### CODE REVIEW

File: state/remote/atlas.go [hunk 1]
```diff
@@ -241,3 +248,72 @@ func (c *AtlasClient) url() *url.URL {
...
+				return conflictHandlingError(err)
+			}
+			return c.Put(buf.Bytes())
+		} else {
+			log.Printf("[DEBUG] States are not equivalent, returning conflict.")
+		}
+	}
+
+	return fmt.Errorf(
+		"Atlas detected a state conflict.\n\nMessage: %s", c.readBody(resp.Body))
```
Comment (sethvargo) [inline state/remote/atlas.go]: "remote state"

File: state/remote/atlas.go [hunk 2]
```diff
@@ -241,3 +248,72 @@ func (c *AtlasClient) url() *url.URL {
...
+// iff for the previous state and the proposed state:
+//
+//   * the serials match
+//   * the parsed states are Equal (semantically equivalent)
+//
+// In other words, in this situation Terraform can override Atlas's detected
+// conflict by asserting that the state it is pushing is indeed correct.
+func (c *AtlasClient) handleConflict(resp *http.Response, state []byte) error {
+	if c.conflictHandlingAttempted && false {
+		log.Printf("[DEBUG] Already attempted conflict resolution; returning conflict.")
```
Comment (ryanuber) [inline state/remote/atlas.go]: This seems unreachable?

Comment (phinze) [inline state/remote/atlas.go]: :see_no_evil: i was test breaking. fixing!

File: state/remote/atlas.go (hunk 1 above)
Comment (phinze) [inline state/remote/atlas.go]: ack

**Comment by ryanuber** [comment]
Left one comment, but this looks sound to me. Great tests too!

**Comment by phinze** [comment]
Addressed comments. :ship:
